File: android/src/main/java/com/mcodex/nitroarchive/domain/ArchivePath.kt
package com.mcodex.nitroarchive.domain

data class ArchivePath(
    val normalized: String,
    val components: List<String>,
    val depth: Int
) {
    companion object {
        private val DRIVE_LETTERS = ('A'..'Z').map { "$it:" } + ('a'..'z').map { "$it:" }

        fun fromRaw(raw: String): ArchivePath {
            require(raw.indexOf('\u0000') == -1) { "Path contains null bytes" }

            val withSlashes = raw.replace("\\", "/")

            require(!withSlashes.startsWith("/")) { "Path starts with /: $raw" }

            require(!withSlashes.startsWith("//")) { "Path is a UNC path: $raw" }

            val upper = withSlashes.uppercase()
            for (drive in DRIVE_LETTERS) {
                require(!upper.startsWith(drive)) { "Path is a Windows drive path: $raw" }
            }

            require(!withSlashes.contains("://")) { "Path contains URI scheme: $raw" }

            val rawComponents = withSlashes.split("/").filter { it.isNotEmpty() }
            val cleaned = mutableListOf<String>()

            for (component in rawComponents) {
                when (component) {
                    "." -> { /* skip */ }
                    ".." -> throw IllegalArgumentException("Path contains '..' component: $raw")
                    else -> cleaned.add(component)
                }
            }

            require(cleaned.isNotEmpty()) { "Path is empty after normalization: $raw" }

            val normalizedPath = cleaned.joinToString("/") + if (raw.endsWith("/")) "/" else ""

            return ArchivePath(
                normalized = normalizedPath,
                components = cleaned.toList(),
                depth = cleaned.size
            )
        }
    }
}

File: ios/Domain/ArchivePath.swift
import Foundation

struct ArchivePath {
  let normalized: String
  let components: [String]
  let depth: Int

  init(raw: String) throws {
    guard !raw.contains("\0") else {
      throw ArchiveDomainError.invalidArgument("Path contains null bytes")
    }

    let replacedBackslash = raw.replacingOccurrences(of: "\\", with: "/")

    guard !replacedBackslash.hasPrefix("/") else {
      throw ArchiveDomainError.absolutePath("Path starts with /: \(raw)")
    }

    guard !replacedBackslash.hasPrefix("//") else {
      throw ArchiveDomainError.pathTraversal("Path is a UNC path: \(raw)")
    }

    for drive in ["A:", "B:", "C:", "D:", "E:", "F:", "G:", "H:", "I:", "J:", "K:", "L:", "M:", "N:", "O:", "P:", "Q:", "R:", "S:", "T:", "U:", "V:", "W:", "X:", "Y:", "Z:"] {
      if replacedBackslash.uppercased().hasPrefix(drive) {
        throw ArchiveDomainError.absolutePath("Path is a Windows drive path: \(raw)")
      }
    }

    guard !replacedBackslash.contains("://") else {
      throw ArchiveDomainError.absolutePath("Path contains URI scheme: \(raw)")
    }

    let rawComponents = replacedBackslash.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
    var cleaned: [String] = []

    for component in rawComponents {
      if component == "." { continue }
      if component == ".." {
        throw ArchiveDomainError.pathTraversal("Path contains '..' component: \(raw)")
      }
      guard !component.isEmpty else { continue }
      cleaned.append(component)
    }

    guard !cleaned.isEmpty else {
      throw ArchiveDomainError.invalidArgument("Path is empty after normalization: \(raw)")
    }

    self.components = cleaned
    self.depth = cleaned.count
    self.normalized = cleaned.joined(separator: "/") + (raw.hasSuffix("/") ? "/" : "")
  }

  func resolved(relativeTo base: String) throws -> String {
    var url = URL(fileURLWithPath: base)
    for component in components {
      url.appendPathComponent(component)
    }
    let resolved = url.path
    guard resolved.hasPrefix(base) else {
      throw ArchiveDomainError.pathTraversal("Resolved path escapes base directory")
    }
    return resolved
  }
}
